replica-node/src/replica_util.rs

```diff
@@ -0,0 +1,264 @@
+        match solana_net_utils::get_cluster_shred_version(entrypoint) {
+            Err(err) => eprintln!("get_cluster_shred_version failed: {}, {}", entrypoint, err),
+            Ok(0) => eprintln!("zero sherd-version from entrypoint: {}", entrypoint),
+            Ok(shred_version) => {
+                info!(
+                    "obtained shred-version {} from {}",
+                    shred_version, entrypoint
+                );
+                return Some(shred_version);
+            }
+        }
+    }
+    None
+}
+
+// Discover the RPC peer node via Gossip and return's ContactInfo
+// And the initial snapshot info: (Slot, Hash)
+// Alternatively, this can be solved via a RPC call instead of using gossip.
+fn get_rpc_peer_node(
+    cluster_info: &ClusterInfo,
+    cluster_entrypoints: &[ContactInfo],
+    expected_shred_version: Option<u16>,
+    peer_pubkey: &Pubkey,
+    snapshot_output_dir: &Path,
+) -> Option<(ContactInfo, Option<(Slot, Hash)>)> {
+    let mut newer_cluster_snapshot_timeout = None;
+    let mut retry_reason = None;
+    loop {
+        sleep(Duration::from_secs(1));
+        info!("Searching for the rpc peer node and latest snapshot information with shred_version {:?}.", expected_shred_version);
+        info!("\n{}", cluster_info.rpc_info_trace());
+
+        let shred_version =
+            expected_shred_version.unwrap_or_else(|| cluster_info.my_shred_version());
+        if shred_version == 0 {
+            let all_zero_shred_versions = cluster_entrypoints.iter().all(|cluster_entrypoint| {
+                cluster_info
+                    .lookup_contact_info_by_gossip_addr(&cluster_entrypoint.gossip)
+                    .map_or(false, |entrypoint| entrypoint.shred_version == 0)
+            });
+
+            if all_zero_shred_versions {
+                eprintln!(
+                    "Entrypoint shred version is zero.  Restart with --expected-shred-version"
+                );
+                exit(1);
+            }
+            info!("Waiting to adopt entrypoint shred version...");
+            continue;
+        }
+
+        info!(
+            "Searching for an RPC service with shred version {}{}...",
+            shred_version,
+            retry_reason
+                .as_ref()
+                .map(|s| format!(" (Retrying: {})", s))
+                .unwrap_or_default()
+        );
+
+        let rpc_peers = cluster_info
+            .all_rpc_peers()
+            .into_iter()
+            .filter(|contact_info| contact_info.shred_version == shred_version)
+            .collect::<Vec<_>>();
+        let rpc_peers_total = rpc_peers.len();
+
+        let rpc_peers_trusted = rpc_peers
+            .iter()
+            .filter(|rpc_peer| &rpc_peer.id == peer_pubkey)
+            .count();
+
+        info!(
+            "Total {} RPC nodes found. {} trusted",
+            rpc_peers_total, rpc_peers_trusted
+        );
+
+        let mut highest_snapshot_hash: Option<(Slot, Hash)> =
```

Comment:
When I was refactoring snapshot_utils code, I was confused when I saw variables called `_hash`, but were really tuples of `(Slot, Hash)`. Would you be willing to rename the variables like this to be `_slot_hash` or something in that vein?

Reply:
renamed to highest_snapshot_info